File: assignment_a1.py
#!/usr/bin/env python3


class StudentGradeManager:
    def __init__(self):
        self.students = []
    
    def calculate_grade(self, marks):
        """Function to calculate grade based on marks using conditionals"""
        if marks >= 90:
            return 'A'
        elif marks >= 75:
            return 'B'
        elif marks >= 50:
            return 'C'
        else:
            return 'F'
    
    def get_remarks(self, grade):
        """Function to get remarks based on grade"""
        remarks_mapping = {
            'A': "Excellent",
            'B': "Very Good", 
            'C': "Needs Improvement",
            'F': "Fail"
        }
        return remarks_mapping.get(grade, "Unknown")
    
    def find_highest_scorer(self):
        """Function to find the highest scorer"""
        if not self.students:
            return None
        
        highest_student = self.students[0]
        for student in self.students:
            if student['marks'] > highest_student['marks']:
                highest_student = student
        return highest_student
    
    def print_result(self, name, marks, remarks=None):
        """
        Method overloading implementation for PrintResult:
        - PrintResult(name, marks) → prints name and grade
        - PrintResult(name, marks, remarks) → prints name, grade, and extra remarks
        """
        grade = self.calculate_grade(marks)
        
        if remarks is None:
            # PrintResult(string name, int marks) → prints name and grade
            print(f"Name: {name}, Marks: {marks}, Grade: {grade}")
        else:
            # PrintResult(string name, int marks, string remarks) → prints name, grade, and extra remarks
            print(f"Name: {name}, Marks: {marks}, Grade: {grade}, Remarks: {remarks}")
    
    def input_students(self):
        """Ask the user for the number of students and input their details"""
        try:
            # Ask the user for the number of students
            num_students = int(input("Enter the number of students: "))
            
            # For each student, input: Name and Marks (0–100)
            for i in range(num_students):
                print(f"\n--- Entering details for Student {i + 1} ---")
                
                # Input student name
                name = input("Enter student name: ").strip()
                
                # Input and validate marks (0-100)
                while True:
                    try:
                        marks = int(input("Enter marks (0-100): "))
                        if 0 <= marks <= 100:
                            break
                        else:
                            print("Error: Please enter marks between 0 and 100.")
                    except ValueError:
                        print("Error: Please enter a valid number.")
                
                # Calculate grade and store student data
                grade = self.calculate_grade(marks)
                student = {
                    'name': name,
                    'marks': marks,
                    'grade': grade
                }
                self.students.append(student)
                
        except ValueError:
            print("Error: Please enter a valid number of students.")
            return False
        return True
    
    def display_all_students(self):
        """Use loops to display each student's Name, Marks, and Grade"""
        print("\n" + "="*60)
        print("               STUDENT GRADE REPORT")
        print("="*60)
        
        # Use loop to display each student's details
        for i, student in enumerate(self.students, 1):
            print(f"\nStudent {i}:")
            remarks = self.get_remarks(student['grade'])
            # Using method overloading with remarks
            self.print_result(student['name'], student['marks'], remarks)
    
    def display_highest_scorer(self):
        """Display the highest scorer using the find_highest_scorer function"""
        highest = self.find_highest_scorer()
        if highest:
            print("\n" + "="*40)
            print("           HIGHEST SCORER")
            print("="*40)
            remarks = self.get_remarks(highest['grade'])
            print(f"🏆 Congratulations to the highest scorer!")
            self.print_result(highest['name'], highest['marks'], remarks)
        else:
            print("No students found.")
    
    def demonstrate_method_overloading(self):
        """Demonstrate method overloading with examples"""
        if not self.students:
            return
            
        print("\n" + "="*50)
        print("        DEMONSTRATING METHOD OVERLOADING")
        print("="*50)
        
        first_student = self.students[0]
        
        print("\n1. Using PrintResult(name, marks) - without remarks:")
        self.print_result(first_student['name'], first_student['marks'])
        
        print("\n2. Using PrintResult(name, marks, remarks) - with remarks:")
        remarks = self.get_remarks(first_student['grade'])
        self.print_result(first_student['name'], first_student['marks'], remarks)
    
    def display_summary_statistics(self):
        """Display additional summary statistics"""
        if not self.students:
            return
            
        print("\n" + "="*40)
        print("         SUMMARY STATISTICS")
        print("="*40)
        
        total_students = len(self.students)
        total_marks = sum(student['marks'] for student in self.students)
        average_marks = total_marks / total_students
        
        # Count grades
        grade_counts = {'A': 0, 'B': 0, 'C': 0, 'F': 0}
        for student in self.students:
            grade_counts[student['grade']] += 1
        
        print(f"Total Students: {total_students}")
        print(f"Average Marks: {average_marks:.2f}")
        print(f"Grade Distribution:")
        for grade, count in grade_counts.items():
            percentage = (count / total_students) * 100
            print(f"  {grade}: {count} students ({percentage:.1f}%)")
    
    def run(self):
        """Main method to run the Student Grade Management System"""
        print("="*60)
        print("     WELCOME TO STUDENT GRADE MANAGEMENT SYSTEM")
        print("                Assignment A1 - Problem 1")
        print("="*60)
        
        # Step 1: Input student data
        if self.input_students():
            # Step 2: Display all students using loops
            self.display_all_students()
            
            # Step 3: Find and display highest scorer
            self.display_highest_scorer()
            
            # Step 4: Demonstrate method overloading
            self.demonstrate_method_overloading()
            
            # Step 5: Display summary statistics
            self.display_summary_statistics()
            
            print("\n" + "="*60)
            print("           THANK YOU FOR USING THE SYSTEM!")
            print("="*60)
        else:
            print("Failed to input student data. Please try again.")

def main():
    """Main function to start the Student Grade Management System"""
    grade_manager = StudentGradeManager()
    grade_manager.run()

if __name__ == "__main__":
    main()

File: Documents/Software-application-development/Assignment_A1.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace StudentGradeManagement
{
    /// <summary>
    /// Assignment A1: Student Grade Management System (20 Marks)
    /// Problem 1: Student Grade Management System
    /// 
    /// Requirements:
    /// 1. Ask the user for the number of students.
    /// 2. For each student, input: Name and Marks (0–100).
    /// 3. Use a function and conditionals to calculate the grade:
    ///    A (90–100), B (75–89), C (50–74), F (<50).
    /// 4. Use loops to display each student's Name, Marks, and Grade.
    /// 5. Use another function to find the highest scorer.
    /// 6. Use another function to get remarks.
    /// 7. Implement method overloading for a function PrintResult.
    /// </summary>
    
    public class Student
    {
        public string Name { get; set; }
        public int Marks { get; set; }
        public char Grade { get; set; }
        
        public Student(string name, int marks, char grade)
        {
            Name = name;
            Marks = marks;
            Grade = grade;
        }
    }
    
    public class StudentGradeManager
    {
        private List<Student> students;
        
        public StudentGradeManager()
        {
            students = new List<Student>();
        }
        
        /// <summary>
        /// Function to calculate grade based on marks using conditionals
        /// </summary>
        /// <param name="marks">Student marks (0-100)</param>
        /// <returns>Grade character (A, B, C, F)</returns>
        public char CalculateGrade(int marks)
        {
            if (marks >= 90)
                return 'A';
            else if (marks >= 75)
                return 'B';
            else if (marks >= 50)
                return 'C';
            else
                return 'F';
        }
        
        /// <summary>
        /// Function to get remarks based on grade
        /// </summary>
        /// <param name="grade">Grade character</param>
        /// <returns>Remarks string</returns>
        public string GetRemarks(char grade)
        {
            switch (grade)
            {
                case 'A':
                    return "Excellent";
                case 'B':
                    return "Very Good";
                case 'C':
                    return "Needs Improvement";
                case 'F':
                    return "Fail";
                default:
                    return "Unknown";
            }
        }
        
        /// <summary>
        /// Function to find the highest scorer
        /// </summary>
        /// <returns>Student with highest marks or null if no students</returns>
        public Student FindHighestScorer()
        {
            if (students.Count == 0)
                return null;
            
            Student highestStudent = students[0];
            foreach (Student student in students)
            {
                if (student.Marks > highestStudent.Marks)
                {
                    highestStudent = student;
                }
            }
            return highestStudent;
        }
        
        /// <summary>
        /// Method overloading for PrintResult
        /// PrintResult(string name, int marks) → prints name and grade
        /// </summary>
        /// <param name="name">Student name</param>
        /// <param name="marks">Student marks</param>
        public void PrintResult(string name, int marks)
        {
            char grade = CalculateGrade(marks);
            Console.WriteLine($"Name: {name}, Marks: {marks}, Grade: {grade}");
        }
        
        /// <summary>
        /// Method overloading for PrintResult
        /// PrintResult(string name, int marks, string remarks) → prints name, grade, and extra remarks
        /// </summary>
        /// <param name="name">Student name</param>
        /// <param name="marks">Student marks</param>
        /// <param name="remarks">Grade remarks</param>
        public void PrintResult(string name, int marks, string remarks)
        {
            char grade = CalculateGrade(marks);
            Console.WriteLine($"Name: {name}, Marks: {marks}, Grade: {grade}, Remarks: {remarks}");
        }
        
        /// <summary>
        /// Ask the user for the number of students and input their details
        /// </summary>
        /// <returns>True if input successful, false otherwise</returns>
        public bool InputStudents()
        {
            try
            {
                // Ask the user for the number of students
                Console.Write("Enter the number of students: ");
                int numStudents = int.Parse(Console.ReadLine());
                
                // For each student, input: Name and Marks (0–100)
                for (int i = 0; i < numStudents; i++)
                {
                    Console.WriteLine($"\n--- Entering details for Student {i + 1} ---");
                    
                    // Input student name
                    Console.Write("Enter student name: ");
                    string name = Console.ReadLine().Trim();
                    
                    // Input and validate marks (0-100)
                    int marks;
                    while (true)
                    {
                        try
                        {
                            Console.Write("Enter marks (0-100): ");
                            marks = int.Parse(Console.ReadLine());
                            if (marks >= 0 && marks <= 100)
                            {
                                break;
                            }
                            else
                            {
                                Console.WriteLine("Error: Please enter marks between 0 and 100.");
                            }
                        }
                        catch (FormatException)
                        {
                            Console.WriteLine("Error: Please enter a valid number.");
                        }
                    }
                    
                    // Calculate grade and store student data
                    char grade = CalculateGrade(marks);
                    Student student = new Student(name, marks, grade);
                    students.Add(student);
                }
                
                return true;
            }
            catch (FormatException)
            {
                Console.WriteLine("Error: Please enter a valid number of students.");
                return false;
            }
        }
        
        /// <summary>
        /// Use loops to display each student's Name, Marks, and Grade
        /// </summary>
        public void DisplayAllStudents()
        {
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine("               STUDENT GRADE REPORT");
            Console.WriteLine(new string('=', 60));
            
            // Use loop to display each student's details
            for (int i = 0; i < students.Count; i++)
            {
                Console.WriteLine($"\nStudent {i + 1}:");
                string remarks = GetRemarks(students[i].Grade);
                // Using method overloading with remarks
                PrintResult(students[i].Name, students[i].Marks, remarks);
            }
        }
        
        /// <summary>
        /// Display the highest scorer using the FindHighestScorer function
        /// </summary>
        public void DisplayHighestScorer()
        {
            Student highest = FindHighestScorer();
            if (highest != null)
            {
                Console.WriteLine("\n" + new string('=', 40));
                Console.WriteLine("           HIGHEST SCORER");
                Console.WriteLine(new string('=', 40));
                string remarks = GetRemarks(highest.Grade);
                Console.WriteLine("🏆 Congratulations to the highest scorer!");
                PrintResult(highest.Name, highest.Marks, remarks);
            }
            else
            {
                Console.WriteLine("No students found.");
            }
        }
        
        /// <summary>
        /// Demonstrate method overloading with examples
        /// </summary>
        public void DemonstrateMethodOverloading()
        {
            if (students.Count == 0)
                return;
                
            Console.WriteLine("\n" + new string('=', 50));
            Console.WriteLine("        DEMONSTRATING METHOD OVERLOADING");
            Console.WriteLine(new string('=', 50));
            
            Student firstStudent = students[0];
            
            Console.WriteLine("\n1. Using PrintResult(name, marks) - without remarks:");
            PrintResult(firstStudent.Name, firstStudent.Marks);
            
            Console.WriteLine("\n2. Using PrintResult(name, marks, remarks) - with remarks:");
            string remarks = GetRemarks(firstStudent.Grade);
            PrintResult(firstStudent.Name, firstStudent.Marks, remarks);
        }
        
        /// <summary>
        /// Display additional summary statistics
        /// </summary>
        public void DisplaySummaryStatistics()
        {
            if (students.Count == 0)
                return;
                
            Console.WriteLine("\n" + new string('=', 40));
            Console.WriteLine("         SUMMARY STATISTICS");
            Console.WriteLine(new string('=', 40));
            
            int totalStudents = students.Count;
            double totalMarks = students.Sum(s => s.Marks);
            double averageMarks = totalMarks / totalStudents;
            
            // Count grades
            var gradeCounts = new Dictionary<char, int>
            {
                {'A', 0}, {'B', 0}, {'C', 0}, {'F', 0}
            };
            
            foreach (Student student in students)
            {
                gradeCounts[student.Grade]++;
            }
            
            Console.WriteLine($"Total Students: {totalStudents}");
            Console.WriteLine($"Average Marks: {averageMarks:F2}");
            Console.WriteLine("Grade Distribution:");
            
            foreach (var gradeCount in gradeCounts)
            {
                double percentage = (double)gradeCount.Value / totalStudents * 100;
                Console.WriteLine($"  {gradeCount.Key}: {gradeCount.Value} students ({percentage:F1}%)");
            }
        }
        
        /// <summary>
        /// Main method to run the Student Grade Management System
        /// </summary>
        public void Run()
        {
            Console.WriteLine(new string('=', 60));
            Console.WriteLine("     WELCOME TO STUDENT GRADE MANAGEMENT SYSTEM");
            Console.WriteLine("                Assignment A1 - Problem 1");
            Console.WriteLine(new string('=', 60));
            
            // Step 1: Input student data
            if (InputStudents())
            {
                // Step 2: Display all students using loops
                DisplayAllStudents();
                
                // Step 3: Find and display highest scorer
                DisplayHighestScorer();
                
                // Step 4: Demonstrate method overloading
                DemonstrateMethodOverloading();
                
                // Step 5: Display summary statistics
                DisplaySummaryStatistics();
                
                Console.WriteLine("\n" + new string('=', 60));
                Console.WriteLine("           THANK YOU FOR USING THE SYSTEM!");
                Console.WriteLine(new string('=', 60));
            }
            else
            {
                Console.WriteLine("Failed to input student data. Please try again.");
            }
        }
    }
    
    /// <summary>
    /// Main program class
    /// </summary>
    class Program
    {
        /// <summary>
        /// Main entry point of the application
        /// </summary>
        /// <param name="args">Command line arguments</param>
        static void Main(string[] args)
        {
            StudentGradeManager gradeManager = new StudentGradeManager();
            gradeManager.Run();
            
            // Keep console open
            Console.WriteLine("\nPress any key to exit...");
            Console.ReadKey();
        }
    }
}
